pandas/core/arrays/categorical.py

```diff
@@ -1229,11 +1229,26 @@ def map(self, mapper):
         Index(['first', 'second', nan], dtype='object')
         """
         new_categories = self.categories.map(mapper)
+
         try:
-            return self.from_codes(self._codes.copy(),
-                                   categories=new_categories,
-                                   ordered=self.ordered)
+            if is_dict_like(mapper):
+                new_value = mapper[np.nan]
+            else:
+                new_value = mapper(np.nan)
+        except (AttributeError, KeyError, TypeError, ValueError):
+            new_value = np.nan
+
+        try:
+            ret = self.from_codes(self._codes.copy(),
+                                  categories=new_categories,
+                                  ordered=self.ordered)
+            if new_value not in ret.categories and any(self._codes == -1):
```

Comment:
This looks incorrect when `new_value` is NA. NA should never be in the categories, so the `add_categories` doesn't make sense.

Reply:
I also don't really like the type instability this creates. Now the output dtype depends on whether there are any missing values present in the column.

Should we instead just say that missing values are not handled in the mapping? That you should instead `.fillna(...).map(mapper)` or `.map(mapper).fillna(...)`, depending on how you want to handle missing values? What does handling missing values in `mapper` give us that `.fillna` doesn't?